### ISSUE TITLE
[3.2.1] File measureLog.xml is not validated by Attestation Service

### ISSUE THREAD

**Issue by mrcdb** [opencit/opencit#61]
I am running Open CIT 3.2.1 on machines with Ubuntu 16.04.2 LTS with TPM 2.0.

Whenever I try to register a new trusted host with Open CIT Attestation Service, I get an IOException and the `catalina,out` log of **mtwilson** shows the following trace:

```
2017-08-08 17:17:32,621 DEBUG [Thread-7] c.i.m.a.i.TAHelper [TAHelper.java:963] Process module [HASH_START] getting extended to [17] has value: e064421772da0cca59cea47801c2ee5e5c2a1758
2017-08-08 17:17:32,626 DEBUG [Thread-7] c.i.m.m.b.HostBO [HostBO.java:137] Unexpected error while retrieving attestation report.
java.io.IOException: java.lang.IllegalArgumentException: No enum constant com.intel.dcsg.cpg.crypto.DigestAlgorithm.
        at com.intel.mtwilson.agent.intel.IntelHostAgent2.getHostAttestationReport(IntelHostAgent2.java:195) ~[mtwilson-intel-hostagent-3.2.1-SNAPSHOT.jar:na]
        at com.intel.mtwilson.ms.business.HostBO$HostAttReport.run(HostBO.java:130) ~[mtwilson-management-3.2.1-SNAPSHOT.jar:na]
Caused by: java.lang.IllegalArgumentException: No enum constant com.intel.dcsg.cpg.crypto.DigestAlgorithm.
        at java.lang.Enum.valueOf(Enum.java:238) ~[na:1.8.0_131]
        at com.intel.dcsg.cpg.crypto.DigestAlgorithm.valueOf(DigestAlgorithm.java:18) ~[mtwilson-util-crypto-3.2.1-SNAPSHOT.jar:na]
        at com.intel.mtwilson.agent.intel.TAHelper.convertHostTpmEventLogEntryToMeasurement(TAHelper.java:1006) ~[mtwilson-intel-hostagent-3.2.1-SNAPSHOT.jar:na]
        at com.intel.mtwilson.agent.intel.TAHelper.verifyQuoteAndGetPcr(TAHelper.java:967) ~[mtwilson-intel-hostagent-3.2.1-SNAPSHOT.jar:na]
        at com.intel.mtwilson.agent.intel.TAHelper.getQuoteInformationForHost(TAHelper.java:593) ~[mtwilson-intel-hostagent-3.2.1-SNAPSHOT.jar:na]
        at com.intel.mtwilson.agent.intel.IntelHostAgent2.getHostAttestationReport(IntelHostAgent2.java:188) ~[mtwilson-intel-hostagent-3.2.1-SNAPSHOT.jar:na]
        ... 1 common frames omitted
```

I have tried to debug the issue, and it seems like `/opt/trustagent/var/measureLog.xml` file does not have the proper syntax. My file looks as follows:

```
<measureLog>
  <txt>
    <txtStatus>3</txtStatus>
    <modules>
      <module>
        <pcrBank></pcrBank>
        <pcrNumber>17</pcrNumber>
        <name>HASH_START</name>
        <value>e064421772da0cca59cea47801c2ee5e5c2a1758</value>
      </module>
      <module>
        <pcrBank></pcrBank>
        <pcrNumber>17</pcrNumber>
        <name>HASH_START</name>
        <value>e064421772da0cca59cea47801c2ee5e5c2a1758</value>
      </module>

[...]
```
As you can see, the `<pcrBank>` tag is empty for PCR 17, and so it is for other registers. Also, some PCRs include `<pcrBank>SHA256</pcrBank>` when they have SHA1 digests instead.

I have tried to manually edit the `measureLog.xml` file by providing the proper values for `<pcrBank>` tags and the issue is solved on the Attestation Service. Btw, the file is recreated at each Trust Agent startup, so this is not a viable workaround IMO.

**Comment by mrcdb** [comment]
Does anyone have the same issue? I just deployed (again) the whole OpenCIT 3.2.1 framework on a OpenStack Mitaka with a TPM 2.0 equipped compute node, and I still get this validation error. The mtwilson DEBUG logging shows that the pcrBank tag is empty...

Just for clarification, find here the full measureLog.xml file with the uncorrect syntax.
[measureLog.xml.txt](https://github.com/opencit/opencit/files/1534805/measureLog.xml.txt)

**Comment by mrcdb** [comment]
More information on the issue...I have investigated where the `measureLog.xml` file is created, and found the `/opt/trustagent/bin/module_analysis.sh` and `/opt/trustagent/bin/module_analysis_da.sh` scripts, which parse the `txt-stat` output to create the XML file.
I have run both the scripts with the `-x` flag to see what the single lines are doing, and realized that the following part in `module_analysis_da.sh` 
```
num="`get_line_number 'Alg: '`"
for parm in $num;do
  if [ $parm -gt $section1EndNum ]; then
    startNum=$endNum
    endNum=$parm
    if [ $startNum -gt $section1EndNum ]; then
      #get all modules' log belong to the alg in startNum line
      AlgId="`$INFILE | sed -n "$startNum p" |  awk -F: '{print $3}'| sed "s/ //g"`"
      for parm2 in $numPcrIndex;do
        if [ $parm2 -gt $startNum -a $parm2 -lt $endNum ];then
          xPcrIndex="`$INFILE | sed -n "$((parm2)) p" |  awk -F: '{print $3}'| sed "s/ //g" | sed "s/\t//g"`"
          xType="`$INFILE | sed -n "$((parm2+1)) p" |  awk -F: '{print $3}'| sed "s/ //g" | sed "s/\t//g"`"
          xDigest="`$INFILE | sed -n "$((parm2+2)) p" |  awk -F: '{print $3}'| sed "s/ //g" | sed "s/\t//g"`"
          xName="`getName $xType`"
          xBank="`getBankById $AlgId`"
          xPcrIndexArray[$modIndex]=$xPcrIndex
          xTypeArray[$modIndex]=$xType
          xDigestArray[$modIndex]=$xDigest
          xNameArray[$modIndex]=$xName
          xBankArray[$modIndex]=$xBank
          let "modIndex++"
        fi
      done
    fi
  fi
done
```
sometimes has the `xBank` variable empty, meaning that `AlgId` is neither `4` (SHA1) or `11` (SHA256). That causes the corresponding `<pcrBank>` parameter to be empty in the `measureLog.xml` file.

In my opinion, the parsing of `txt-stat` does not work well with the current version of TBOOT that I have installed in the Trust Agent (1.9.5).

**Comment by mrcdb** [comment]
Hi all,
I just updated the `opencit/opencit-trustagent` code, starting from tag 3.2.1, to generate a correct `measureLog.xml` file in a TPM 2.0 environment with TBOOT 1.9.5. The changes are limited to a specific script, which is executed at each boot by the trust agent. See here https://github.com/mrcdb/opencit-trustagent/commit/9512e7ab88e84bf13b595e0be0eb52d68f294c48 for details.

Also, how can we contribute to the project codebase via pull requests? I have tried to fork the trust agent repository but I cannot PR back in the base code.
